### DISCUSSION TITLE
How do I delete a ScaledObject and all its assocated objects

### DISCUSSION THREAD

**Discussion by georgestevens99** [kedacore/keda#2301] [Q&A / Need Help] [answered]
I am a software developer and architect who has been using K8s since June 2021.  Thus I do not have a full grasp of K8s, but I know enough to develop app services, deploy them to K8s, and debug them.  

I recently successfully used KEDA to scale out my app services based on the number of items in an Azure Service Bus Topic.  That learning experience was OK since the KEDA documentation (and other articles as well) provided a clear path to achieving the result I wanted.  Thanks!

Now I want to delete the KEDA ScaledObject, the related HPA object, plus the various other K8s objects that got installed with them so I can maintain a "cleak cluster"  **I can find absolutely nothing in your documentation about how to delete these things**!

Please do 2 things to help:
1. Respond to this request with clear instructions on how to delete these things.  PS I do not have a clue what a "finalizer" is.  I'd like to delete these via a command (kubectl?) rather than by editing K8s objects in a running cluster so that I can make a shell script file to run for deletion in the future.

2. Modify your KEDA documentation to have an easy-to-find section on how to delete ScaledObjects.

Thanks,
George

**Comment by tomkerkhove** [comment, marked as answer]
There isn't much to document as you just have to delete your ScaledObject, that's all there is 🤷‍♂️

```
kubectl delete scaledobject/<name>
```

All other rest will be managed for you.

If you want to remove KEDA, you can just use `helm uninstall` if you have used Helm to install it in the first place.

Hope this helps.

**Reply by deadlysyn** [reply]
not working, stuck on finalizer.

```console
kubectl patch scaledobject/<name> --type json --patch='[ { "op": "remove", "path": "/metadata/finalizers" } ]'
kubectl delete scaledobject/<name>
```

**Reply by JorTurFer** [reply]
Maybe this link is useful for you: https://keda.sh/docs/latest/deploy/#uninstall

**Comment by georgestevens99** [comment]
Thanks for the quick reply!  I am glad to see deleting a ScaledObject is so
simple!

**Comment by crosbymichael1** [comment]
Anytime I try to delete my scaled objects they DONT delete because of the finalizers. How can we delete a scaled Object without having to hack a way to remove the finalizer from the Scaled Object? @tomkerkhove 

Also what is the purpose of having the finalizer on the scaled Object other than just not allowing someone to delete the object?

This doesnt work for me:

kubectl delete scaledobject/name

**Reply by tomkerkhove** [reply]
What version are you running? Finalizers are there to do cleanup such as removing the HPA, unless @zroubalik corrects me.

**Reply by crosbymichael1** [reply]
Im currently deploying KEDA into a EKS cluster via a Helm Chart:

chart         = "kedacore/keda"
chart_version = "2.3.2"

Each of our customer's resources (deployments, services) are deployed to their own namespace where we create ScaledObjects in each namespace as well. The problem is when we want to delete a customer, we cant delete the namespace because of the finalizers in the scaledObjects. We've had to resort to adding logic in our jenkins job to remove the finalizer from the scaledObjects before we can delete the namespace and it doesnt feel like thats something we should have to do. How do people normally clean up their scaledObjects they no longer need anymore?

Heres the code we added to remove the finalizers before deleting the namespace:

```
for i in $(k get scaledobject | awk '{ print $1 }' | awk 'NR!=1 {print}'); 
do kubectl patch scaledobject $i -p '{"metadata":{"finalizers":null}}' --type=merge
done
```

**Reply by crosbymichael1** [reply]
@tomkerkhove

**Reply by tomkerkhove** [reply]
It's not recommended to remove the finalizers or your HPAs will stay around unless the whole namespace is removed but frankly I don't think there is much we can improve here.

What do you think @zroubalik / @JorTurFer?

**Reply by JorTurFer** [reply]
hey!
It doesn't make sense. The finalizers are a mechanism in k8s to ensure that the deletions are done properly. KEDA adds its own finalizer to the ScaledObjects to ensure that it's notified when they are deleted to clean up the resources. 
These blocks could happen if KEDA operator pod is down, in that case the finalizer blocks the deletion, but I assume that you have one KEDA instance in the whole cluster and not one per namespace, right?
Could you share the operator logs when you try to delete the ScaledObject @crosbymichael1 ?

**Reply by adrianparks** [reply]
To reiterate the OP's statement: the KEDA documentation is very good indeed; the install process was easy to do and understand. But when I follow the uninstall section of https://keda.sh/docs/2.6/deploy/, my ScaledObjects (and associated CRD) are not removed (understandably, as Helm didn't create them). Probably all that is needed is that the uninstall docs mention that all SOs should be removed before running `helm uninstall`, or you will end up having to do the workaround @crosbymichael1 details above? That worked fine for me.

**Reply by zroubalik** [reply]
good idea! are you willing to contribute this?

**Reply by adrianparks** [reply]
Seems fair! Let me go read up on the contributing guidelines and I will submit a PR.

**Reply by zroubalik** [reply]
There is `Suggest change` button on top of https://keda.sh/docs/2.6/deploy/ page. So opening a PR against this repo is enough.

**Reply by crosbymichael1** [reply]
@JorTurFer Sorry to rehash this but Im still having issues and adrianparks comment didnt resolve my query. 

When I run:
`kubectl delete scaledobject alert-builder -n michael-pz`

Then it gets hung up on the below message:
`scaledobject.keda.sh "alert-builder" deleted`

I search through the KEDA Operator logs during this and no logs were generated for issuing that k delete scaledobject command and the SO didnt delete I still see it in the cluster.

**Reply by JorTurFer** [reply]
are you using namespaced KEDA Operator?

**Reply by crosbymichael1** [reply]
The keda operator is in a namespace called 'infra' thats shared with some other pods. Im wondering if me attempting to delete the scaled object doesnt show up in the keda operator logs at all because something on the kubernetes side is blocking the delete attempt at all since it has a finalizer.

**Reply by crosbymichael1** [reply]
Here is what one of my scaled objects looks like, you can notice the finalizer is also defined.

```
kubectl describe scaledobjects alert-builder

Name:         alert-builder
Namespace:    demo2
Labels:       app.kubernetes.io/managed-by=Helm
              scaledObjectName=alert-builder
Annotations:  meta.helm.sh/release-name: alert-builder-<redacted> 
              meta.helm.sh/release-namespace: <redacted> 
API Version:  keda.sh/v1alpha1
Kind:         ScaledObject
Metadata:
  Creation Timestamp:  2021-11-23T07:27:36Z
  Finalizers:
    finalizer.keda.sh
  Generation:  2
  Managed Fields:
    API Version:  keda.sh/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:annotations:
          .:
          f:meta.helm.sh/release-name:
          f:meta.helm.sh/release-namespace:
        f:labels:
          .:
          f:app.kubernetes.io/managed-by:
      f:spec:
        .:
        f:maxReplicaCount:
        f:minReplicaCount:
        f:pollingInterval:
        f:scaleTargetRef:
          .:
          f:name:
        f:triggers:
    Manager:      Go-http-client
    Operation:    Update
    Time:         2021-11-23T07:27:36Z
    API Version:  keda.sh/v1alpha1
    Fields Type:  FieldsV1
    fieldsV1:
      f:metadata:
        f:finalizers:
          .:
          v:"finalizer.keda.sh":
        f:labels:
          f:scaledObjectName:
      f:status:
        .:
        f:conditions:
        f:externalMetricNames:
        f:lastActiveTime:
        f:originalReplicaCount:
        f:scaleTargetGVKR:
          .:
          f:group:
          f:kind:
          f:resource:
          f:version:
        f:scaleTargetKind:
    Manager:         keda
    Operation:       Update
    Time:            2021-11-23T08:06:48Z
  Resource Version:  1175065984
  UID:               5e98724f-e7de-464c-a456-6ef3b6c4980d
Spec:
  Max Replica Count:  5
  Min Replica Count:  2
  Polling Interval:   60
  Scale Target Ref:
    Name:  alert-builder
  Triggers:
    Authentication Ref:
      Name:  keda-trigger-kafka-auth
    Metadata:
      Allow Idle Consumers:  true
      Bootstrap Servers:    <redacted> 
      Consumer Group:        <redacted> 
      Lag Threshold:         2000
      Offset Reset Policy:   latest
      Topic:                 <redacted> 
    Type:                    kafka
Status:
  Conditions:
    Message:  ScaledObject is defined correctly and is ready for scaling
    Reason:   ScaledObjectReady
    Status:   True
    Type:     Ready
    Message:  Scaling is not performed because triggers are not active
    Reason:   ScalerNotActive
    Status:   False
    Type:     Active
  External Metric Names:
    <redacted> 
  Last Active Time:        2022-07-11T23:00:03Z
  Original Replica Count:  2
  Scale Target GVKR:
    Group:            apps
    Kind:             Deployment
    Resource:         deployments
    Version:          v1
  Scale Target Kind:  apps/v1.Deployment
Events:               <none>
```

**Reply by JorTurFer** [reply]
Could you try setting the loglevel to debug in the operator? certainly it's weird

**Reply by zroubalik** [reply]
The finalizer is owned by KEDA, so you should check that it is working correctly. There might be some misconfiguration since you are not used the default `keda` namespace.
So k8s is waiting till KEDA Operator finalizes that resource

**Reply by crosbymichael1** [reply]
So then what is the thing that is supposed to send a message to the keda operator when I issue that kubectl delete scaledobject command? Does keda-operator watch all its scaledobjects and then see when one is tried to be deleted, does the scaledobject or k8s send a request to keda-operator to delete it?

**Reply by JorTurFer** [reply]
k8s notifies the deletion to KEDA Operator. That's why you cannot remove them if KEDA Operator pod is not present because it needs to "approve" it, otherwise k8s blocks de deletion

**Reply by crosbymichael1** [reply]
How does k8s know where the keda-operator pod is? I dont see anything defined in the scaledobject that would define that. Also thank you for your help, I appreciate it.

**Reply by JorTurFer** [reply]
KEDA uses the [operator framework](https://github.com/operator-framework/operator-sdk) and it provides some "magic". KEDA registers the finalizer and the framework doesn the magic under the hood

**Reply by crosbymichael1** [reply]
I think i'm understanding a little more, I think by us manually removing those finalizers and then deleting the scaled object, KEDA thinks the scaledobjects are still there as I see many of these errors below for namespaces that no longer exist.


```
{"level":"error","ts":1657643700.3194726,"logger":"scalehandler","msg":"Error getting scalers","object":{"apiVersion":"keda.sh/v1alpha1","kind":"ScaledObject","namespace":"itg-9bb62d37","name":"datumwriter-service"},"error":"Deployment.apps \"datumwriter-service\" not found","stacktrace":"github.com/kedacore/keda/v2/pkg/scaling.(*scaleHandler).startScaleLoop\n\t/workspace/pkg/scaling/scale_handler.go:156"}
```

I can try to put the keda-operator in a KEDA namespace but im wondering if theres something at the app level thats hardcoded to look in the KEDA ns, or why I couldnt put it in any namespace name I chose?

**Reply by crosbymichael1** [reply]
So before I moved Keda into a different ns called KEDA, I tried to delete a SO in my preprod env as KEDA is also in the infra ns there.  

I was able to delete the SO immediately.
```
k delete scaledobjects cloudwatch-integration
scaledobject.keda.sh "cloudwatch-integration" deleted
```

And I also see the corresponding logs in the Keda-operator for the reconciliation of the deleted SO:
```
{"level":"info","ts":1657644809.8109486,"logger":"controller.scaledobject","msg":"Reconciling ScaledObject","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"cloudwatch-integration","namespace":"mcp-12898"}

{"level":"info","ts":1657644809.8267329,"logger":"controller.scaledobject","msg":"Successfully finalized ScaledObject","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"cloudwatch-integration","namespace":"mcp-12898"}
```

**Reply by JorTurFer** [reply]
I don't think that the problem it's the names pace itself tbh. Does it happen also with a new ScaledObject or only with olds?

**Reply by JorTurFer** [reply]
Restarting operator pod, can you see any error? I'm suspicing that an RBAC issue could be there (because I assume that the traffic from KEDA to k8s api it's not blocked by any reason, right?)

**Reply by crosbymichael1** [reply]
I restarted the pod now and I can see that its reconciling all scaled objects so far but dont see an error. Another interesting thing I noticed is that in my preprod cluster when I created a SO the KEDA-Operator logs showed the below recognizing that a SO was created, but in prod when I did the same I didnt see any log mentioning 'Reconciling ScaledObject'

```
{"level":"info","ts":1657644843.408781,"logger":"controller.scaledobject","msg":"Reconciling ScaledObject","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"cloudwatch-integration","namespace":"keda-test-2"}

{"level":"info","ts":1657644843.4088023,"logger":"controller.scaledobject","msg":"Adding Finalizer for the ScaledObject","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"cloudwatch-integration","namespace":"keda-test-2"}

{"level":"info","ts":1657644843.5486264,"logger":"controller.scaledobject","msg":"Detected resource targeted for scaling","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"cloudwatch-integration","namespace":"keda-test-2","resource":"apps/v1.Deployment","name":"cloudwatch-integration"}

{"level":"info","ts":1657644843.5487475,"logger":"controller.scaledobject","msg":"Creating a new HPA","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"cloudwatch-integration","namespace":"keda-test-2","HPA.Namespace":"keda-test-2","HPA.Name":"keda-hpa-cloudwatch-integration"}

{"level":"info","ts":1657644843.5641296,"logger":"controller.scaledobject","msg":"Initializing Scaling logic according to ScaledObject Specification","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"cloudwatch-integration","namespace":"keda-test-2"}

{"level":"info","ts":1657644843.5920813,"logger":"controller.scaledobject","msg":"Reconciling ScaledObject","reconciler group":"keda.sh","reconciler kind":"ScaledObject","name":"cloudwatch-integration","namespace":"keda-test-2"}
```

**Reply by crosbymichael1** [reply]
Also after restarting the KEDA-Operator pod in prod I dont see any more errors saying that scaledobject cant be found (for the old SO that were deleted)
